nixos/modules/services/mail/postsrsd.nix

```diff
@@ -107,29 +90,42 @@ in
       postsrsd.gid = config.ids.gids.postsrsd;
     };
 
+    systemd.services.postsrsd-generate-secrets = {
+      path = [ pkgs.coreutils ];
+      script = ''
+        if [ -e "${cfg.secretsFile}" ]; then
+          echo "Secrets file exists. Nothing to do!"
+        else
+          echo "WARNING: secrets file not found, autogenerating!"
+          DIR="$(dirname "${cfg.secretsFile}")"
+          install -m 750 -o ${cfg.user} -g ${cfg.group} -d "$DIR"
+          install -m 600 -o ${cfg.user} -g ${cfg.group} <(dd if=/dev/random bs=18 count=1 | base64) "${cfg.secretsFile}"
+        fi
+      '';
+      serviceConfig = {
+        Type = "oneshot";
+      };
+    };
+
     systemd.services.postsrsd = {
       description = "PostSRSd SRS rewriting server";
-      after = [ "network.target" ];
+      after = [
+        "network.target"
+        "postsrsd-generate-secrets.service"
+      ];
       before = [ "postfix.service" ];
       wantedBy = [ "multi-user.target" ];
-
-      path = [ pkgs.coreutils ];
+      requires = [ "postsrsd-generate-secrets.service" ];
+      confinement.enable = true;
```

Comment:
```suggestion
      confinement.enable = true;
      
      preScript = ''
        if [ ! -e "${cfg.secretsFile}" ]; then
          echo "Secrets file not found, autogenerating..."
          DIR="$(dirname "${cfg.secretsFile}")"
          install -m 750 -o ${cfg.user} -g ${cfg.group} -d "$DIR"
          install -m 600 -o ${cfg.user} -g ${cfg.group} <(dd if=/dev/random bs=18 count=1 | base64) "${cfg.secretsFile}"
        fi
      '';
```

Reply:
Can you explain how this works with `confinement.enable = true;`?  The reason I split the generation into a separate, unconfined service was because this (confined) service can't write to arbitrary locations on the filesystem.

Reply:
I have never used confinement and don't know how it interacts with the service at all.

My thought was, that it would be simpler to just have one systemd unit without any dependencies. If you feel comfortable with confinement feel free to resolve this.

Reply:
I'd like to leave it this way. I get the sense that upstream would like us to run the service unprivileged (see [docs](https://github.com/roehling/postsrsd/blob/a20e0b646c87f13cc013dd8aa8c221283a5a06b1/doc/postsrsd.conf#L159-L177) and the [default setuid/setgid](https://github.com/roehling/postsrsd/blob/65141d580f2c215bb700a7d8cbadc9ef829f6b71/src/main.c#L66)).